community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/RWLock.java

```diff
@@ -313,62 +352,95 @@ else if ( we.lockType == LockType.READ )
+     * In that case lock will not be acquired and false will be return as result of acquisition
      *
      * @throws DeadlockDetectedException
      *             if a deadlock is detected
+     * @return true is lock was acquired, false otherwise
      */
-    synchronized void acquireWriteLock( Object tx ) throws DeadlockDetectedException
+    synchronized boolean acquireWriteLock( Object tx ) throws DeadlockDetectedException
     {
         TxLockElement tle = getOrCreateLockElement( tx );
 
+        LockRequest lockRequest = null;
+        // used to track do we need to add lock request to a waiting queue or we still have it there
+        boolean addLockRequest = true;
         try
         {
-            tle.movedOn = false;
-
-            boolean shouldAddWait = true;
+            tle.incrementRequests();
             Thread currentThread = currentThread();
 
-            while ( totalWriteCount > tle.writeCount || totalReadCount > tle.readCount )
+            while ( !tle.isTerminated() && (totalWriteCount > tle.writeCount || totalReadCount > tle.readCount) )
             {
                 ragManager.checkWaitOn( this, tx );
 
-                if (shouldAddWait)
+
+                if ( addLockRequest )
                 {
-                    waitingThreadList.addFirst( new WaitElement( tle, WRITE, currentThread) );
+                    lockRequest = new LockRequest( tle, WRITE, currentThread );
+                    waitingThreadList.addFirst( lockRequest );
                 }
 
                 try
                 {
                     wait();
-                    shouldAddWait = false;
+                    addLockRequest = false;
                 }
                 catch ( InterruptedException e )
                 {
                     interrupted();
-
-                    shouldAddWait = true;
+                    addLockRequest = true;
                 }
                 ragManager.stopWaitOn( this, tx );
             }
 
-            registerWriteLockAcquired( tx, tle );
+            if ( !tle.isTerminated() )
+            {
+                registerWriteLockAcquired( tx, tle );
+                return true;
+            } else {
+                // in case if lock element was interrupted and it was never register before
+                // we need to clean it from lock element map
+                // if it was register before it will be cleaned up during standard lock release call
+                if ( tle.requests == 1 && tle.isFree() )
+                {
+                    txLockElementMap.remove( tx );
+                }
+                return false;
+            }
         }
         finally
         {
+            cleanupWaitingListRequests( lockRequest, tle, addLockRequest );
+            // for cases when spurious wake up was the reason why we waked up, but also there
+            // was an interruption as described at 17.2 just clearing interruption flag
+            interrupted();
```

Comment:
Could we end up swallowing interrupts here, that we shouldn't have? It looks racy. I don't think we're using interrupts for signalling anywhere else in our code base, but you never know what can happen in embedded, even though other parts of our code base (such as the transaction log) isn't interruption safe.